glean_parser/lint.py

```diff
@@ -0,0 +1,194 @@
+            f"'{common_prefix}'. Remove prefixes and (possibly) rename category."
+        )
+
+
+def check_unit_in_name(metric):
+    """
+    The metric name ends in a unit.
+    """
+    TIME_UNIT_ABBREV = {
+        "nanosecond": "ns",
+        "microsecond": "us",
+        "millisecond": "ms",
+        "second": "s",
+        "minute": "m",
+        "hour": "h",
+        "day": "d",
+    }
+
+    MEMORY_UNIT_ABBREV = {
+        "byte": "b",
+        "kilobyte": "kb",
+        "megabyte": "mb",
+        "gigabyte": "gb",
+    }
+
+    name_words = _split_words(metric.name)
+    unit_in_name = name_words[-1]
+
+    if hasattr(metric, "time_unit"):
+        if (
+            unit_in_name == TIME_UNIT_ABBREV.get(metric.time_unit.name)
+            or unit_in_name == metric.time_unit.name
+        ):
+            yield (
+                f"Suffix '{unit_in_name}' is redundant with time_unit. "
+                "Only include time_unit."
+            )
+        elif (
+            unit_in_name in TIME_UNIT_ABBREV.keys()
+            or unit_in_name in TIME_UNIT_ABBREV.values()
+        ):
+            yield (
+                f"Suffix '{unit_in_name}' doesn't match time_unit. "
+                "Only include time_unit."
+            )
+
+    elif hasattr(metric, "memory_unit"):
+        if (
+            unit_in_name == MEMORY_UNIT_ABBREV.get(metric.memory_unit.name)
+            or unit_in_name == metric.memory_unit.name
+        ):
+            yield (
+                f"Suffix '{unit_in_name}' is redundant with memory_unit. "
+                "Only include memory_unit."
+            )
+        elif (
+            unit_in_name in MEMORY_UNIT_ABBREV.keys()
+            or unit_in_name in MEMORY_UNIT_ABBREV.values()
+        ):
+            yield (
+                f"Suffix '{unit_in_name}' doesn't match memory_unit. "
+                "Only include memory_unit."
+            )
+
+    elif hasattr(metric, "unit"):
+        if unit_in_name == metric.unit:
+            yield (
+                f"Suffix '{unit_in_name}' is redundant with unit param. "
+                "Only include unit."
+            )
+
+
+def check_category_generic(category_name, metrics):
+    """
+    The category name is too generic.
+    """
+    GENERIC_CATEGORIES = ["metrics", "events"]
+
+    if category_name in GENERIC_CATEGORIES:
+        yield f"Category '{category_name}' is too generic."
+
+
+CATEGORY_CHECKS = {
+    "COMMON_PREFIX": check_common_prefix,
+    "CATEGORY_GENERIC": check_category_generic,
+}
+
+
+INDIVIDUAL_CHECKS = {"UNIT_IN_NAME": check_unit_in_name}
+
+
+def lint_metrics(objs, file=sys.stderr):
+    """
+    Performs glinter checks on a set of metrics objects.
+
+    :param objs: Tree of metric objects, as returns by `parser.parse_objects`.
+    :param file: The stream to write errors to.
+    :returns: List of nits.
+    """
+    nits = []
+    for (category_name, metrics) in sorted(list(objs.items())):
```

Comment:
Do we need to filter out objects coming from `pings.yaml`? Do the same rules apply there too?

Reply:
Good catch.  Let's just skip pings for now.